net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+static int libfab_conn_init(struct m0_fab__ep *ep, struct m0_fab__tm *ma,
+			    struct m0_fab__buf *fbp);
+static int libfab_conn_accept(struct m0_fab__ep *ep, struct m0_fab__tm *tm,
+			      struct fi_info *info);
+static int libfab_fab_ep_find(struct m0_fab__tm *tm, struct m0_fab__ep_name *en,
+			      const char *name, struct m0_fab__ep **ep);
+static void libfab_ep_pton(struct m0_fab__ep_name *name, uint64_t *out);
+static void libfab_ep_ntop(uint64_t netaddr, struct m0_fab__ep_name *name);
+static void libfab_txep_event_check(struct m0_fab__ep *txep,
+				    struct m0_fab__active_ep *aep,
+				    struct m0_fab__tm *tm);
+static void libfab_rxep_comp_read(struct fid_cq *cq, struct m0_fab__ep *ep);
+static void libfab_txep_comp_read(struct fid_cq *cq);
+static int libfab_txep_init(struct m0_fab__active_ep *aep,
+			    struct m0_fab__tm *tm);
+static int libfab_waitfd_bind(struct fid* fid, struct m0_fab__tm *tm);
+static inline struct m0_fab__active_ep *libfab_aep_get(struct m0_fab__ep *ep);
+static int libfab_bulk_op(struct m0_fab__active_ep *ep, struct m0_fab__buf *fb);
+static inline bool libfab_is_verbs(struct m0_fab__tm *tm);
+static int libfab_bulklist_add(struct m0_fab__tm *tm, struct m0_fab__buf *fb,
+				struct m0_fab__active_ep *aep);
+static uint32_t libfab_wr_cnt_get(struct m0_fab__buf *fb);
+static void libfab_bulk_buf_process(struct m0_fab__tm *tm);
+
+/* libfab init and fini() : initialized in motr init */
+M0_INTERNAL int m0_net_libfab_init(void)
+{
+	int result = 0;
+
+	m0_net_xprt_register(&m0_net_libfab_xprt);
+	m0_net_xprt_default_set(&m0_net_libfab_xprt);
+	return M0_RC(result);
+}
+
+M0_INTERNAL void m0_net_libfab_fini(void)
+{
+	m0_net_xprt_deregister(&m0_net_libfab_xprt);
+}
+
+/**
+ * This function decodes the lnet format address and extracts the ip address and
+ * port number from it.
+ * This is also used to allocate unique transfer machine identifiers for LNet
+ * network addresses with wildcard transfer machine identifier (like
+ * "192.168.96.128@tcp1:12345:31:*").
+ */
+static int libfab_ep_addr_decode_lnet(const char *name, char *node,
+				      size_t nodeSize, char *port,
+				      size_t portSize)
+{
+	char     *at = strchr(name, '@');
+	int       nr;
+	unsigned  pid;
+	unsigned  portal;
+	unsigned  portnum;
+	unsigned  tmid;
+	char      *lp = "127.0.0.1";
+
+	if (strncmp(name, "0@lo", 4) == 0) {
+		M0_PRE(nodeSize >= ((strlen(lp)+1)) );
+		memcpy(node, lp, (strlen(lp)+1));
+	} else {
+		if (at == NULL || at - name >= nodeSize)
+			return M0_ERR(-EPROTO);
+
+		M0_PRE(nodeSize >= (at-name)+1);
+		memcpy(node, name, at - name);
+	}
+	if ((at = strchr(at, ':')) == NULL) /* Skip 'tcp...:' bit. */
+		return M0_ERR(-EPROTO);
+	nr = sscanf(at + 1, "%u:%u:%u", &pid, &portal, &tmid);
+	if (nr != 3) {
+		nr = sscanf(at + 1, "%u:%u:*", &pid, &portal);
+		if (nr != 2)
+			return M0_ERR(-EPROTO);
+		for (nr = 0; nr < ARRAY_SIZE(fab_autotm); ++nr) {
```

Comment:
I don't like to re-use "nr" as iterator here.
Let's define a new variable "i" and use "i".

Reply:
Corrected